### PR TITLE
[InstCombine] Make fptrunc combine use intersection of fast math flags

### PR DESCRIPTION
**Pull request by john-brawn-arm**
These combines involve swapping the fptrunc with its operand, and using the intersection of fast math flags is the safest option as e.g. if we have (fptrunc (fneg ninf x)) then (fneg ninf (fptrunc x)) will not be correct as if x is a not within the range of the destination type the result of (fptrunc x) will be inf.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 1]
```diff
@@ -1847,15 +1847,16 @@ Instruction *InstCombinerImpl::visitFPTrunc(FPTruncInst &FPT) {
   Value *X;
   Instruction *Op = dyn_cast<Instruction>(FPT.getOperand(0));
   if (Op && Op->hasOneUse()) {
-    // FIXME: The FMF should propagate from the fptrunc, not the source op.
     IRBuilder<>::FastMathFlagGuard FMFG(Builder);
+    FastMathFlags FMF = FPT.getFastMathFlags();
     if (isa<FPMathOperator>(Op))
-      Builder.setFastMathFlags(Op->getFastMathFlags());
+      FMF &= Op->getFastMathFlags();
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:1853]: Description says union but this is intersection.Also dyn_cast to FPMathOperator and query flags from that, this is effectively isa + cast

File: llvm/test/Transforms/InstCombine/fpcast.ll [hunk 2]
```diff
@@ -78,6 +89,28 @@ define half @test4-fast(float %a) {
...
+; CHECK-LABEL: @test4-mixed-fast-1(
+; CHECK-NEXT:    [[TMP1:%.*]] = fptrunc float [[A:%.*]] to half
+; CHECK-NEXT:    [[C:%.*]] = fneg half [[TMP1]]
+; CHECK-NEXT:    ret half [[C]]
+;
+  %b = fsub float -0.0, %a
+  %c = fptrunc fast float %b to half
+  ret half %c
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fpcast.ll:108]: Test a vector case, and with a subset of flags

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp (hunk 1 above)
Comment (john-brawn-arm) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp]: Yes, I meant intersection.

**Review by arsenm — APPROVED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**